conformance/utils/kubernetes/helpers.go

```diff
@@ -275,6 +275,11 @@ func GatewayAndHTTPRoutesMustBeAccepted(t *testing.T, c client.Client, timeoutCo
 						Status: metav1.ConditionTrue,
 						Reason: string(v1beta1.RouteReasonAccepted),
 					},
+					{
```

Comment:
When reading this helper more carefully, it looks like the Gateway conditions are not checked here (they are actually checked in no `HTTPRoute` test). The helper checks the `ParentConditions`, which by definition (from docs) are:

```
// Conditions describes the status of the route with respect to the Gateway.
// Note that the route's availability is also subject to the Gateway's own
// status conditions and listener status.
// ...
```

For this reason, I'd remove this `ResolvedRefs` condition from here (sorry for suggesting it :)) and seek other's opinion on checking/not checking gateway conditions in `HTTPRoute` tests. Is it something you deem we should take care of? Or just performing this check in the Gateway tests is enough?
I feel we should add such a check in all tests that create gateways, but it is not a really strong opinion.

@shaneutt @robscott @sunjayBhatia @michaelbeaumont @LiorLieberman @arkodg

Reply:
Good catch @mlavacca! This led me to question what we actually _require_ here. At first I looked at the API Spec, and the only references I could find required it to be set when it was negative. Fortunately this is clarified in the [GEP that prompted this](https://gateway-api.sigs.k8s.io/geps/gep-1364/#should-conditions-always-be-added), and it appears that `ResolvedRefs` should indeed always be set. 

My next question is _where_ it should be set. For some reason, `ResolvedRefs` is a `Listener` condition and not a `Gateway` condition:

https://github.com/kubernetes-sigs/gateway-api/blob/f638d3fe75a69354b4a15b6c9e9e475c4e7d8062/apis/v1beta1/gateway_types.go#L833-L849

I _think_ that makes sense because 100% of the object references in a Gateway are defined within a Listener today. But it could get awfully noisy to have to specify a positive value for each listener in a Gateway here. I wonder if it would make more sense just to require that a top level positive condition is set on the Gateway, and if there are problems, they'll be surfaced on individual listeners as well?

In any case, to do a universal check for this condition, we'd likely want to add it somewhere like `GatewayAndHTTPRoutesMustBeAccepted` (along with some renaming) since that function appears to be pretty broadly used in our conformance tests.

Open to any/all ideas here, but particularly curious what @youngnick thinks since he's been so instrumental in defining how our status + conditions work.

/cc @youngnick

Reply:
seems like this carries over to Accepted and Programmed conditions as well, if I'm not mistaken we have inconsistent checks for those Listener conditions in tests as well

Also agreed `GatewayAndHTTPRoutesMustBeAccepted` could use a rename/rework to include some element of/call `GatewayStatusMustHaveListeners` (since as it is today it just checks on a Gateway for the latest generation on top-level conditions and for an address to be assigned)

Reply:
Thinking in terms of how a user finds out what is wrong with their Gateway today in Contour (without adding a having a top-level summary condition besides what exists), say in the case of an invalid TLS certificateRef:
- See Programmed=False with reason "ListenersNotValid"
- Drill down into Listeners to see what is wrong
- See a Listener is Programmed=False, and also ResolvedRefs=False with reason "InvalidCertificateRef"

Seems like there is enough of a flow/hierarchy to get the info you need currently

Right now we don't set ResolvedRefs on every Listener but we could, it would add some more info and clutter, but it feels like Accepted/Programmed are good enough summary conditions at the top-level and ListenerStatus level

Interested in others thoughts

some example status:
```
  status:
    addresses:
    - type: IPAddress
      value: 172.18.255.200
    conditions:
    - lastTransitionTime: "2023-08-10T19:41:11Z"
      message: Gateway is accepted
      observedGeneration: 3
      reason: Accepted
      status: "True"
      type: Accepted
    - lastTransitionTime: "2023-08-10T19:41:11Z"
      message: Listeners are not valid
      observedGeneration: 3
      reason: ListenersNotValid
      status: "False"
      type: Programmed
    listeners:
    - attachedRoutes: 0
      conditions:
      - lastTransitionTime: "2023-08-10T19:41:11Z"
        message: 'Spec.VirtualHost.TLS.CertificateRefs "foo" referent is invalid:
          Secret not found'
        observedGeneration: 3
        reason: InvalidCertificateRef
        status: "False"
        type: ResolvedRefs
      - lastTransitionTime: "2023-08-10T19:41:11Z"
        message: Invalid listener, see other listener conditions for details
        observedGeneration: 3
        reason: Invalid
        status: "False"
        type: Programmed
      - lastTransitionTime: "2023-08-10T19:41:11Z"
        message: Listener accepted
        observedGeneration: 3
        reason: Accepted
        status: "True"
        type: Accepted
      name: site-invalid-https
      supportedKinds:
      - group: gateway.networking.k8s.io
        kind: HTTPRoute
      - group: gateway.networking.k8s.io
        kind: GRPCRoute
```

Reply:
So to my understanding, we need to check inside `GatewayAndHTTPRoutesMustBeAccepted` if every listener has the `ListenerConditionResolvedRefs` set to true.

I think the easiest way to do it is to add a new function (say we call it `GatewayListenersMustHaveCondition`) similar to `HTTPRouteMustHaveCondition`, which checks if every listener of the specified gateway has a certain condition (in our case - `ListenerConditionResolvedRefs`).

This new function will be called inside `GatewayAndHTTPRoutesMustBeAccepted`.

Let me know what you think about it @mlavacca @robscott @sunjayBhatia

Reply:
@robscott

Reply:
> So to my understanding, we need to check inside `GatewayAndHTTPRoutesMustBeAccepted` if every listener has the `ListenerConditionResolvedRefs` set to true.
> 
> I think the easiest way to do it is to add a new function (say we call it `GatewayListenersMustHaveCondition`) similar to `HTTPRouteMustHaveCondition`, which checks if every listener of the specified gateway has a certain condition (in our case - `ListenerConditionResolvedRefs`).
> 
> This new function will be called inside `GatewayAndHTTPRoutesMustBeAccepted`.

Yeah, it sounds good to me.

Reply:
The requested change has been added, so I resolved this conversation.
Please reopen if necessary.